rfc/src/rfcs/0008-function-contracts.md

```diff
@@ -0,0 +1,623 @@
+support needed for many unbounded problems, such as ["decreases"
+measures](https://diffblue.github.io/cbmc/contracts-decreases.html) and
+inductive lemmas (for instance in [ACSL](https://frama-c.com/download/acsl.pdf)
+section 2.6.3 "inductive predicates").
+
+## User Experience
+
+Function contract provide a verifiable way to specify function behavior. In
+addition the specified behavior can subsequently be used as an abstraction of
+the functions behavior at call sites.
+
+The lifecycle of a contract is split roughly into three phases: specification,
+verification and stubbing. Which we will explore on this simple example:
+
+```rs
+fn my_div(dividend: u32, divisor: u32) -> u32 {
+  dividend / divisor
+}
+```
+
+1. In the first phase we **specify** the approximation. Kani provides two new
+   annotations: `requires` (preconditions) to describe the expectations this
+   function has as to the calling context and `ensures` (postconditions) which
+   approximates function outputs in terms of function inputs.
+
+   ```rs
+   #[kani::requires(divisor != 0)]
+   #[kani::ensures(result <= dividend)]
+   fn my_div(dividend: u32, divisor: u32) -> u32 {
+     dividend / divisor
+   }
+   ```
+  
+   `requires` here indicates this function expects its `divisor` input to never
+   be 0, or it will not execute correctly (i.e. panic).
+
+   `ensures` puts a bound on the output, relative to the `dividend` input.
+
+   Conditions in contracts are plain Rust expressions which can reference the
+   function arguments and, in case of `ensures`, the result of the function as a
+   special `result`[^result-naming] variable. Syntactically Kani supports any
+   Rust expression, including function calls, defining types etc. However they
+   must be side-effect free[^side-effects] or Kani will throw a compile error.
+
+   Multiple `requires` and `ensures` clauses are allowed on the same function,
+   they are implicitly logically conjoined.
+
+   [^result-naming]: See [open questions](#open-questions) for a discussion
+       about naming of the result variable.
+
+2. Next Kani must make sure that the approximation we specified actually holds
+   by **checking** the contract against the implementation. This is in contrast
+   to ["stubbing"][stubbing], where the approximation is blindly trusted.
+
+   The contract must always overapproximate the function to guarantee soundness.
+
+   To facilitate the check Kani needs a suitable environment to verify the
+   function in. For this proposal the environment must be provided by us (the
+   users). See [future possibilities](#future-possibilities) for a discussion
+   about the arising soundness issues and their remedies.
+
+   We provide the checking environment for our contract with a special new
+   `proof_for_contract` harness.
+
+   ```rs
+   #[kani::proof_for_contract(my_div)]
+   fn my_div_harness() {
+     my_div(kani::any(), kani::any())
+   }
+   ```
```

Comment:
In theory yes, we should be able to check contracts for any harness that contains a contracts clause. In practice, we can't yet. We could make automatic harness generation a requirement for stabilization (at least for now) and the harness for contracts a temporary work around? What do you think @adpaco-aws?

Reply:
Because heaps are complicated. This is what Remi was working on and hasn't yet finished. Basically it's hard to create arbitrary heap datastructures that you know will exercise all behaviors of the function you want to check.

Reply:
That's a technical limitation which doesn't seem to be mentioned anywhere. However, it looks like we're working around it with sacrifices to UX. Wouldn't it be better to reduce the scope and go for the optimal UX?

In other words, if we can detect the cases where we need arbitrary heap data structures, we can refuse contract checking/replacement. That way, UX doesn't need to change when the work there is completed, we can just remove the mitigation (detection).

Reply:
Resolved offline.

Reply:
I've added this discussion to the "Rational" section.

Imo the limitations this UX would impose is unreasonable. I'd rather enable users to be productive and gradually improve their UX.